### PR TITLE
bin: Prevent free on signal handler

### PR DESCRIPTION
**Pull request by nokute78**
free() is not signal safe function. https://man7.org/linux/man-pages/man7/signal-safety.7.html
It means we can't call flb_stop/flb_destroy on signal handler.

This patch is to simplify the signal handler.
* Signal handler only changes `exit_signal` variable.
* If `exit_signal` is not 0 , it breaks sleeping loop and invokes current signal handler function from main.
* SIGSEGV and SIGCONT are handled same way.
----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [N/A] Example configuration file for the change
- [X] Debug log output from testing the change

- [] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [N/A] Documentation required for this feature

## Debug Output

SIGINT:
```
$ bin/fluent-bit -i cpu -o stdout
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/06/09 22:40:10] [ info] [engine] started (pid=32052)
[2021/06/09 22:40:10] [ info] [storage] version=1.1.1, initializing...
[2021/06/09 22:40:10] [ info] [storage] in-memory
[2021/06/09 22:40:10] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/06/09 22:40:10] [ info] [sp] stream processor started
^C[2021/06/09 22:40:10] [engine] caught signal (SIGINT)
[2021/06/09 22:40:10] [ info] [input] pausing cpu.0
[2021/06/09 22:40:10] [ warn] [engine] service will stop in 5 seconds
[2021/06/09 22:40:14] [ info] [engine] service stopped
```

SIGSEGV:
```
$ bin/fluent-bit -i cpu -o null
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/06/12 15:04:55] [ info] [engine] started (pid=36504)
[2021/06/12 15:04:55] [ info] [storage] version=1.1.1, initializing...
[2021/06/12 15:04:55] [ info] [storage] in-memory
[2021/06/12 15:04:55] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/06/12 15:04:55] [ info] [sp] stream processor started
[2021/06/12 15:05:07] [engine] caught signal (SIGSEGV)
#0  0x7f9ddc5dd3bf      in  ???() at ???:0
#1  0x7f9ddc5e3046      in  ???() at ???:0
#2  0x7f9ddc5e2f7d      in  ???() at ???:0
#3  0x564203497288      in  flb_main() at src/fluent-bit.c:1205
#4  0x5642034972f3      in  main() at src/fluent-bit.c:1221
#5  0x7f9ddc5240b2      in  ???() at ???:0
#6  0x5642034943ed      in  ???() at ???:0
#7  0xffffffffffffffff  in  ???() at ???:0
Aborted (core dumped)
```

SIGQUIT:
```
d$ bin/fluent-bit -i cpu -o null
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/06/09 22:44:42] [ info] [engine] started (pid=36570)
[2021/06/09 22:44:42] [ info] [storage] version=1.1.1, initializing...
[2021/06/09 22:44:42] [ info] [storage] in-memory
[2021/06/09 22:44:42] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/06/09 22:44:42] [ info] [sp] stream processor started
[2021/06/09 22:44:55] [engine] caught signal (SIGQUIT)
[2021/06/09 22:44:55] [ info] [input] pausing cpu.0
[2021/06/09 22:44:55] [ warn] [engine] service will stop in 5 seconds
[2021/06/09 22:44:59] [ info] [engine] service stopped
```

SIGCONT:
```
$ bin/fluent-bit -i cpu -o null
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/06/09 22:45:29] [ info] [engine] started (pid=36573)
[2021/06/09 22:45:29] [ info] [storage] version=1.1.1, initializing...
[2021/06/09 22:45:29] [ info] [storage] in-memory
[2021/06/09 22:45:29] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/06/09 22:45:29] [ info] [sp] stream processor started
[2021/06/09 22:45:34] [engine] caught signal (SIGCONT)
[2021/06/09 22:45:34] Fluent Bit Dump

===== Input =====
cpu.0 (cpu)
│
├─ status
│  └─ overlimit     : no
│     ├─ mem size   : 125b (125 bytes)
│     └─ mem limit  : 0b (0 bytes)
│
├─ tasks
│  ├─ total tasks   : 0
│  ├─ new           : 0
│  ├─ running       : 0
│  └─ size          : 0b (0 bytes)
│
└─ chunks
   └─ total chunks  : 1
      ├─ up chunks  : 1
      ├─ down chunks: 0
      └─ busy chunks: 0
         ├─ size    : 0b (0 bytes)
         └─ size err: 0


===== Storage Layer =====
total chunks     : 1
├─ mem chunks    : 1
└─ fs chunks     : 0
   ├─ up         : 0
   └─ down       : 0
```

SIGTERM:
```
$ bin/fluent-bit -i cpu -o null
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/06/09 22:46:04] [ info] [engine] started (pid=36576)
[2021/06/09 22:46:04] [ info] [storage] version=1.1.1, initializing...
[2021/06/09 22:46:04] [ info] [storage] in-memory
[2021/06/09 22:46:04] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/06/09 22:46:04] [ info] [sp] stream processor started
[2021/06/09 22:46:15] [engine] caught signal (SIGTERM)
[2021/06/09 22:46:15] [ info] [input] pausing cpu.0
[2021/06/09 22:46:15] [ warn] [engine] service will stop in 5 seconds
[2021/06/09 22:46:20] [ info] [engine] service stopped
```

SIGHUP:
```
$ bin/fluent-bit -i cpu -o null
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/06/09 22:46:40] [ info] [engine] started (pid=36580)
[2021/06/09 22:46:40] [ info] [storage] version=1.1.1, initializing...
[2021/06/09 22:46:40] [ info] [storage] in-memory
[2021/06/09 22:46:40] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/06/09 22:46:40] [ info] [sp] stream processor started
[2021/06/09 22:46:50] [engine] caught signal (SIGHUP)
[2021/06/09 22:46:50] [ info] [input] pausing cpu.0
[2021/06/09 22:46:50] [ warn] [engine] service will stop in 5 seconds
[2021/06/09 22:46:54] [ info] [engine] service stopped
```
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by nokute78** [comment]
Note: Ctrl+C twice is ignored.
Current master exits promptly, on the other hand, this patch causes waiting grace period.

File: src/flb_lib.c [hunk 1]
```diff
@@ -664,14 +664,6 @@ int flb_start(flb_ctx_t *ctx)
     return 0;
 }
 
-int flb_loop(flb_ctx_t *ctx)
```
Comment (edsiper) [inline src/flb_lib.c:667]: is it necessary to remove this function ?, note that the usage of this function can be outside of the scope of the caller  fluent-bit.c, e.g: a custom application using Fluent Bit in library mode where is not associated to signals

**Review by edsiper — CHANGES_REQUESTED** [review]

File: src/flb_lib.c (hunk 1 above)
Comment (nokute78) [reply src/flb_lib.c:667]: Oh, is it exposed as a library API ?
I won't remove such APIs.

Comment (edsiper) [reply src/flb_lib.c:667]: yeah, that should not be removed

Comment (nokute78) [reply src/flb_lib.c]: Sure. I removed the commit not to delete `flb_loop`

**Comment by nokute78** [comment]
I updated PR to modify only fluent-bit.c.

**Pull request merged**